solutions/advanced/Ligatures.mdx

```diff
@@ -0,0 +1,272 @@
+<CPPSection>
+
+```cpp
+#pragma GCC target("avx2")
+#include <immintrin.h>
+#include <iostream>
+#include <string>
+#include <vector>
+using namespace std;
+
+#define rep(i, from, to) for (int i = from; i < (to); ++i)
+#define trav(a, x) for (auto &a : x)
+#define all(x) x.begin(), x.end()
+#define sz(x) (int)(x).size()
+typedef long long ll;
+typedef pair<int, int> pii;
+typedef vector<int> vi;
+
+// Sets up the environment for using SIMD operations with AVX2 instructions
+const int SIMD_BITS = 256;
+const int MAXQ = 100'000;
+
+typedef __m256i mi;
+
+// Sets a specific bit in a 256-bit vector (mi type) at the given index (ind)
+void set_bit(mi &m, size_t ind) {
+	union {
+		mi m;
+		uint32_t ar[SIMD_BITS / 32];
+	} u;
+	u.m = m;
+	u.ar[ind / 32] = (uint32_t)(u.ar[ind / 32] | (1 << (ind % 32)));
+	m = u.m;
+}
+
+// Prints the binary representation of a 256-bit vector (mi type)
+void printm(mi m) {
+	union {
+		mi m;
+		uint32_t ar[SIMD_BITS / 32];
+	} u;
+	u.m = m;
+	rep(i, 0, SIMD_BITS) { cout << ((u.ar[i / 32] >> (i % 32)) & 1); }
+	cout << endl;
+}
+
+mi active[MAXQ / SIMD_BITS + 1][26 * 26 + 1]{};  // ~26*26 * MAXQ / 8 bytes
+
+int main() {
+	cin.sync_with_stdio(false);   // Disable synchronization with C stdio
+	cin.exceptions(cin.failbit);  // Enable exceptions on input failure
+	int N, Q, K;
+	cin >> N >> Q >> K;
+	string str, s;
+	cin >> str;
+	N--;
+	int padN = (N + 31) & -32;        // Pad N to the nearest multiple of 32
+	int qchunks = Q / SIMD_BITS + 1;  // Calculate number of query chunks
+	vector<short> pairs(padN);        // Vector to store character pairs
+	rep(i, 0, N) {
+		int a = str[i] - 'a';
+		int b = str[i + 1] - 'a';
+		pairs[i] = (short)((26 * a + b) * sizeof(mi));
+	}
+	rep(i, N, padN) {
+		pairs[i] = 26 * 26 * sizeof(mi);
+	}  // Fill remaining pairs with a default value
+
+	// Process each query and set the corresponding bits in the active array
+	rep(i, 0, Q) {
+		cin >> s;
+		rep(j, 0, K) {
+			int a = s[2 * j] - 'a';
+			int b = s[2 * j + 1] - 'a';
+			set_bit(active[i / SIMD_BITS][26 * a + b], i % SIMD_BITS);
+		}
+	}
+
+	vi res(qchunks * SIMD_BITS);
+	rep(k, 0, qchunks) {
+		char *act = (char *)active[k];
+		int i = 0;
+		mi zero = _mm256_setzero_si256();
+		mi acc0 = zero, acc1 = zero, acc2 = zero, acc3 = zero, acc4 = zero;
+		mi accs[30]{}, ts[30]{};
+		mi picked = zero;
+		mi carry;
+#define SET0                                                                   \
+	carry = picked = _mm256_andnot_si256(picked, *(mi *)(act + pairs[i++]))
+#define ADD2(t, acc)                                                           \
+	{                                                                          \
+		mi x = t ^ carry, y = (x & acc) | (t & carry);                         \
+		acc = x ^ acc, carry = y;                                              \
+	}  // (carry, acc) = (t + acc + carry)
+#define ADD(ind) ADD2(t##ind, acc##ind)
+
+		// Loop to process the pairs and accumulate the results
+		while (i < padN) {
```

Comment:
remove the loop unrolling - the solution still runs fast enough without it, and it adds 100 lines to the code length

it should suffice to remove acc0 ... acc4 and start `ind` at 0

Reply:
after this you should also be able to remove `padN`